[WIP] Improvements to the Lua map generator

CelticMinstrel opened the pull request.

CelticMinstrel commented:
Anyone have any idea why the latest commit doesn't work? There's two problems with it:

1. If I use `map:set_tile` like the comment says it should, I get an assertion error saying that the map's width is broken (there are more or fewer terrains in some row than the map width).
2. With the current implementation, the road is patchy and disconnected.

gfgtdf commented:
@CelticMinstrel you have changed the milestone to 1.14.1 but since it has changes that add new syntaxes/features this cannot be done during a stable release. We could eigher make it before 1.14 or at 1.15 .

CelticMinstrel commented:
@gfgtdf – Ultimately I wanted to have a multiplayer cave map which basically shows off the new cave generator features, and I believe the road cost thing is basically needed for that. I think the patchy and disconnected aspect of the road is probably something about order of operations (as if the road is being added correctly, but then something else is overwriting it later).

CelticMinstrel commented:
Okay... there are still some things I want to do (mainly adding user config option on the cave map), but I think this could be mergeable at this point.

The road algorithm does seem like it could use some improvement, mind you - the road I'm getting is pretty patchy. Tweaks to the scenario WML for the cave map would also be welcome (maybe @sigurdfdragon would have some ideas on improving it?).

sigurdfdragon commented:
I may be able to take a look at this on Thursday.

sigurdfdragon commented:
@CelticMinstrel The things that may fix the patchiness are:
1. There isn't a [road_cost] entry for all types of terrain that might be encounter. At least Sm (muddy swamp) and Rb^Fetd are listed as clear terrains in the generator, and there's no road cost them. Try adding those and looking for others. The gist is to make sure there is a  [road_cost] tag for anything the generator may encounter from point A to B.

2. Try more standard road costs for various terrains (ie, 10, 20, 25, 30 etc) Use 50 as the highest. Use 2 as the lowest amount. I don't know why / don't remember why mainline doesn't use 1 for MIN_ROAD_COST macros, but there's probably a reason.

CelticMinstrel commented:
Thanks for the tips. I'll try applying these suggestions, probably tomorrow.

CelticMinstrel commented:
Those tips seem to have worked, thanks! I think this can now be merged; I do still want to extend it to allow up to 9 players (or at least 8; 9 might be a bit hard), and to allow some configuration of parameters, but aside from that it all seems to be fully-functional, and the map it generates doesn't seem to be too terribly boring either. At most it could use maybe a few tweaks, maybe making the lake bigger, adding a river, or something like that.

CelticMinstrel commented:
...okay, I spoke a little too soon. There's this weird bug whereby the map (sometimes?) doesn't generate until you click "Regenerate". That is, at first you see nothing and it says the map is 0x0. Click "Regenerate", and it all works.

gfgtdf commented:
please make the return value of the lua user_config optional.

CelticMinstrel commented:
Okay, so basically all working now, I guess. There's still some room for improvement, mind you; if you make the lake large enough, you'll get villages dotted through it, for example. I also got twice as many castles as players, for some reason.

CelticMinstrel commented:
In other words, unless @sigurdfdragon or others want to suggest some tweaks, this can be merged whenever.

sigurdfdragon commented:
I'll be able to take another look at this on Wednesday.

gfgtdf commented:
just to be sure, this doesn't change behaviour of the 2 mainline scneraios where the cave random map generator is used, right?

CelticMinstrel commented:
It shouldn't, since it's _only_ adding additional capabilities to the generator and using them in a brand-new MP scenario. Still, it wouldn't hurt to check, just to be sure...

(That said, I'd argue that we _should_ consider using some of these new capabilities in the two mainline scenarios. With these upgrades it could be possible to get rid of all those `SCATTER_EMBELLISHMENTS` macro calls in the HTTT scenario, for example. However, that's outside of the scope of this PR.)

sigurdfdragon commented:
I found the following bugs:
1. The number of villages produced on HttT 17 & SoF 4 is consistently lower than on current master. (ie, around 6-7 on SoF 4, whereas on master approx 15-35) The villages that are produced are at the bottom of the map, and the count thins out when you go up. This is most obvious on HttT 17
2. The double keeps seem to only start occurring when the settings box is opened, regardless of wether or not anything is changed. Once it happens, it's sticks, until the create game screen has been exited and re-enterted.
3. <illegal tile in map (2 Ko) '2 Ko'> (or similar) occurs occasionally on regenerating. The smaller the map & higher number of player seems to make it more likely to occur? My guess is that it might be trying to put a keep on top of an existing keep? (fixing double keep probably fixes this)
4. 'Roads between castles' checkbox does not appear to work, as there are still roads when the checkbox is removed.
5. Not sure if it is intentional or not, but the roads generated look to be a width of 3, whereas roads on the default map generator are typically a width of 1 hex.


Lake villages - A way of dealing with them could be to add a prestart event to convert villages surrounded by water into merfolk villages.

CelticMinstrel commented:
The roads were intentionally wide, with `width=2`... not sure why they're 3 wide though when I'm asking for 2 wide. I'll look into these issues, maybe this weekend or early next week.

Vultraz commented:
Do you still have issues with the settings dialog?

sigurdfdragon commented:
I should be able to check the settings issue post https://github.com/wesnoth/wesnoth/commit/86f026281bf13b9e03345726cc6d74b627b3d7b2

by the end of the week

Vultraz commented:
Ok, I'm just asking since @CelticMinstrel mentioned some problems with the dialog and if those problems have to do with checkboxes then 5a05d7e might fix them.

CelticMinstrel commented:
It's also possible that the `roads_toggled` function isn't correctly implemented, or the `ignore` key isn't correctly honoured - there are several possible points of failure.

sigurdfdragon commented:
Did a post 86f0262 test. It doesn't seem to help with any of the issues I mentioned above

gfgtdf commented:
i wonder whether we should enable `wensoth.get_varibale` in lua map generators so that the author can generate maps based on the variables set by the scenarios. Just a random idea though.

CelticMinstrel commented:
That sounds like it'd be a lot of work, given that the WML context doesn't even exist yet when the generator runs. I suppose a way to fetch just the custom options might be useful, though.

Vultraz commented:
*pokes @CelticMinstrel to do something with this*

CelticMinstrel commented:
Pretty sure I won't have time this weekend; any time I do have for Wesnoth will most likely be focused on the deprecation PR. I do want to get the material changes in this PR into master for 1.14, though, so if that's still an option after 1.13.11, maybe then? Obviously that would omit the new MP scenario (as that would violate the string freeze), but everything else could be merged

Vultraz commented:
Postponed until 1.14.1.

gfgtdf commented:
as noted eariler this cannot be done in a stable release, if it won't go into 1.14 it must be 1.15

CelticMinstrel commented:
Huh. This shockingly has zero conflicts. Maybe I can actually test and merge it real quick before dealing with the Lua game map API...

virtualghetto commented:
`local prev_x, prev_y` on line [201](https://github.com/wesnoth/wesnoth/blob/10cd9d6d74dc4c8f251294cd67fde14046efda46/data/lua/cave_map_generator.lua#L201) need to be moved above line 198 for the bridges to work. Currently it will give a row of castles instead because prev_x and prev_y are always nil.

CelticMinstrel commented:
I'm not quite sure I understand? They're not _always_ nil, since they're set at the end of the iteration over that location set… unless the location set always has only one element…?

virtualghetto commented:
Yes, that's it. If the passage width is 1, you'd get a nil prev_x and prev_y.
It might be better for the prev_x and prev_y to be based on the path instead of build_chamber.
I haven't tested width > 1 honestly, but for width = 1 I got a row of castles instead of bridges.

virtualghetto commented:
This is what I used for testing. It's not the map generator, but it copied over the road generation code and logic. Uncomment line 123 to see the row of castles over the river.
[_main.cfg.txt](https://github.com/wesnoth/wesnoth/files/5743497/_main.cfg.txt)

virtualghetto commented:
I did more tests on this one.
This time I used the given cave_map_generator and mapgen_helper from the branch.

I had trouble getting the cave_map_generator to work at first, wesnoth kept quitting on me.
Turns out the cave_map_generator needed the following line at the top to work.
`local helper = wesnoth.require "lua/helper.lua"`

I created a map with Wwg as its default terrain_wall and 2 chambers at the corners.
I created a single passage between the 2 chambers using the road costs defined in Random_Scenario_Cave.cfg.

Bridge cost is defined as:
```
        [road_cost]
                terrain=Wwg
                cost=100
                convert_to_bridge=Wwg^Bw|,Wwg^Bw/,Wwg^Bw\
                convert_to=Ur
        [/road_cost]
```
Passage width is 1 and `local prev_x, prev_y` position is unchanged
![width_1_old](https://user-images.githubusercontent.com/10203479/103386438-bb512680-4af6-11eb-9c44-552625db00d0.png)

Passage width is 2 and `local prev_x, prev_y` position is unchanged
![width_2_old](https://user-images.githubusercontent.com/10203479/103386481-e63b7a80-4af6-11eb-8f44-1e511dfe64b9.png)

Passage width is 1 and `local prev_x, prev_y` moved outside the for loop.
![width_1_new](https://user-images.githubusercontent.com/10203479/103386496-f5bac380-4af6-11eb-883d-ca9a778a6cfd.png)

Passage width is 2 and `local prev_x, prev_y`  moved outside the for loop.
![width_2_new](https://user-images.githubusercontent.com/10203479/103386516-11be6500-4af7-11eb-980a-ceeab08119d4.png)

virtualghetto commented:
Hi again!
I did more tests with this version of cave_map_generator and found some more issues to consider.

The relative_to values are all off by 1. This could result in unplayable maps or generation errors if everything was relative to x,y=1,1.
Some map sizes could throw a lua error because of the division by 2.

A possible fix could be:
```
                if chamber.relative_to == "top-right" then
                        x = map.w - 1 - x
                elseif chamber.relative_to == "bottom-right" then
                        x = map.w - 1 - x
                        y = map.h - 1 - y
                elseif chamber.relative_to == "bottom-left" then
                        y = map.h - 1 - y
                elseif chamber.relative_to == "top-middle" then
                        x = math.floor((map.w - 1 - x) / 2)
                elseif chamber.relative_to == "bottom-middle" then
                        x = math.floor((map.w - 1 - x) / 2)
                        y = map.h - 1 - y
                elseif chamber.relative_to == "middle-left" then
                        y = math.floor((map.h - 1 - y) / 2)
                elseif chamber.relative_to == "middle-right" then
                        y = math.floor((map.h - 1 - y) / 2)
                        x = map.w - 1 - x
                elseif chamber.relative_to == "center" then
                        x = math.floor((map.w - 1 - x) / 2)
                        y = math.floor((map.h - 1 - y) / 2)
                end -- Default is "top-left" which means no adjustments needed
```

Using the same path calculation to build the passage and road at the same time could result in taking the highest road_cost terrain especially for passage width = 1.
So you could end up with a road that uses cost=1000 convert_to the entire way!

A possible fix would be calling find_path twice, once for the passage to tunnel its way with clear_tile, and another for laying out the road with place_road.

I also think roads having width > 1 don't look right.

```
			local calc = function(x, y)
				local res = 1.0
				local tile = map:get_tile(x, y)
				if tile == params.terrain_wall then
					res = laziness
				end
				if windiness > 1 then
					res = res * random(windiness)
				end
				return res
			end
			local path = wesnoth.find_path(
				v.start_x, v.start_y, v.dest_x, v.dest_y, calc, map.w, map.h)

			for i, loc in ipairs(path) do
				local locs_set = LS.create()
				build_chamber(loc[1], loc[2], locs_set, width, jagged)
				for x,y in locs_set:stable_iter() do
					local r = 1000
					local ter_to_place
					if v.data.place_villages then r = random(1000) end
					if r <= params.village_density then
						ter_to_place = v.data.terrain_village or params.terrain_village
					else
						ter_to_place = v.data.terrain_clear or params.terrain_clear
					end
					clear_tile(x, y, ter_to_place)
				end
			end

			local calc = function(x, y)
				local res = 1.0
				local tile = map:get_tile(x, y)
				res = v.costs[tile] or 1.0
				if windiness > 1 then
					res = res * random(windiness)
				end
				return res
			end
			local path = wesnoth.find_path(
				v.start_x, v.start_y, v.dest_x, v.dest_y, calc, map.w, map.h)

			local prev_x, prev_y
			for i, loc in ipairs(path) do
				local locs_set = LS.create()
				insert_locs(loc[1], loc[2], locs_set)
				for x,y in locs_set:stable_iter() do
					ter_to_place = v.data.terrain_clear or params.terrain_clear
					place_road(x, y, prev_x, prev_y, v.roads, ter_to_place)
					prev_x, prev_y = x, y
				end
			end
```

cooljeanius commented:
be sure to document what changes UMC authors will be required to make when using the Lua map generator after this gets merged, and possibly add `wmllint` rules to automatically make those changes for people when possible

Wedge009 commented:
Is this only 'postponed' because of 1.14 vs 1.15? 1.16 is imminent, so I suppose this has missed that window as well...

CelticMinstrel commented:
Yeah, pretty much. I hope to get back to it this year though.

Wedge009 commented:
...and we're now on 1.16 vs 1.17. (:

(Sorry, was just reviewing 'postponed' issues again.)

CelticMinstrel commented:
I _would_ like to add this to the roadmap for 1.17 but I'm a) not sure where to put it and b) not sure if I'll even get to it.

Wedge009 commented:
I had the impression that Pentarctagon thought there wasn't much logged in the road-map, so I suppose it doesn't really matter where you put it? As far back as possible, I suppose, if you're uncertain on time-frame.

CelticMinstrel commented:
My current todo list for this PR:

* [ ] [Issues noticed in HttT](https://github.com/wesnoth/wesnoth/pull/993#issuecomment-326171320)
* [ ] [Bridges issue](https://github.com/wesnoth/wesnoth/pull/993#issuecomment-747907212)
* [ ] [Evaluate the big chunk of code at the end here](https://github.com/wesnoth/wesnoth/pull/993#issuecomment-753252907)
* [ ] #5838
* [ ] #5370

CelticMinstrel commented:
Updated todo list (copied from https://github.com/wesnoth/wesnoth/pull/993#issuecomment-1189686300 and deleted completed tasks):

* [ ]  Villages dotting the lake should be converted to water villages
* [ ]  [Various issues, including in HttT](https://github.com/wesnoth/wesnoth/pull/993#issuecomment-326171320)
* * [ ]  Village distribution issue
* * [ ]  Occasional "illegal tile (2 Ko)"?
* * [ ]  Roads are too wide
* [ ]  [Bridges issue](https://github.com/wesnoth/wesnoth/pull/993#issuecomment-747907212)
* [ ]  [wesnoth.unit_types and wesnoth.terrain_types are now available in map generation scripts #5838](https://github.com/wesnoth/wesnoth/pull/5838)

cooljeanius commented:
How are those merge conflicts looking?

CelticMinstrel commented:
Glancing at them in the web resolve UI, they mostly don't look too bad. The schema validator conflict might be a bit tricky, but it might also turn out to be trivial at a closer look.

Wedge009 commented:
I notice there's a few conflicts to resolve before this can be merged.

What was broken about the HttT Sceptre of Fire scenario? I tried building this branch as-is, but I keep getting compilation failures on `mapgen_lua_kernel.cpp`.

babaissarkar commented:
I have rebased this branch on top of `master`, shouldn't be any more conflicts.

babaissarkar commented on src/scripting/mapgen_lua_kernel.cpp:
```diff
@@ -302,9 +307,15 @@ void mapgen_lua_kernel::run_generator(const char * prog, const config & generato
 }
 
-void mapgen_lua_kernel::user_config(const char * prog, const config & generator)
+void mapgen_lua_kernel::user_config(const char * prog, config & generator)
```
```suggestion
void mapgen_lua_kernel::user_config(const char* prog, config& generator)
```

CelticMinstrel commented:
@Wedge009 I think [this](https://github.com/wesnoth/wesnoth/pull/993#issuecomment-326171320) is the original comment about the issue. Many changes were made since then though.

Vultraz commented:
Looks like a missing const on that config parameter.

Wedge009 commented:
I might be doing something wrong, but I'm still not able to build this.
